File: src/main/java/demo/bfims/Entities/Inventory/PublicationItem.java
package demo.bfims.Entities.Inventory;

import demo.bfims.Enums.ItemType;
import demo.bfims.Enums.PublicationItemFormat;
import demo.bfims.Enums.PublicationItemStatus;
import demo.bfims.Enums.PublicationItemType;
import jakarta.persistence.*;

@Entity
public abstract class PublicationItem extends Item {
    @Enumerated(EnumType.STRING)
    private PublicationItemFormat format;
    @Enumerated(EnumType.STRING)
    private PublicationItemStatus status;
    @ManyToOne(cascade = {CascadeType.MERGE, CascadeType.PERSIST, CascadeType.REFRESH},  fetch = FetchType.LAZY)
    @JoinColumn(name = "publication_id")
    Publication publication;
    private Double purchasePrice;
    private Double rentalRate;
    private String edition;
    @Enumerated(EnumType.STRING)
    private PublicationItemType publicationItemType;


    public PublicationItem() {
        this.setItemType(ItemType.PUBLICATION_ITEM);
    }

    public PublicationItemStatus getStatus() {
        return status;
    }

    public void setStatus(PublicationItemStatus status) {
        this.status = status;
    }

    public PublicationItemFormat getFormat() {
        return format;
    }

    public void setFormat(PublicationItemFormat format) {
        this.format = format;
    }

    public Publication getPublication() {
        return publication;
    }

    public void setPublication(Publication publication) {
        this.publication = publication;
        // assuming a new publication item is being added
        if(this.getPublication() != null)
            this.publication.increaseQuantity();
    }

    public Double getPurchasePrice() {
        return purchasePrice;
    }

    public void setPurchasePrice(Double purchasePrice) {
        this.purchasePrice = purchasePrice;
    }

    public Double getRentalRate() {
        return rentalRate;
    }

    public void setRentalRate(Double rentalRate) {
        this.rentalRate = rentalRate;
    }

    public String getEdition() {
        return edition;
    }

    public void setEdition(String edition) {
        this.edition = edition;
    }

    public PublicationItemType getPublicationItemType() {
        return publicationItemType;
    }

    public void setPublicationItemType(PublicationItemType publicationItemType) {
        this.publicationItemType = publicationItemType;
    }

    @Override
    public String toString() {
        return "PublicationItem{" +
                "format=" + format +
                ", status=" + status +
                ", publication=" + publication +
                ", purchasePrice=" + purchasePrice +
                ", rentalRate=" + rentalRate +
                ", edition='" + edition + '\'' +
                ", publicationItemType=" + publicationItemType +
                "} " + super.toString();
    }
}

File: frontend/src/app/DTOs/Inventory/PublicationItem.ts
import {Item} from './Item';
import {Publication} from './Publication';
import {ItemType} from '../../Enums/ItemType'
import {PublicationItemType} from '../../Enums/PublicationItemType';
import {PublicationItemStatus} from '../../Enums/PublicationItemStatus';
import {PublicationItemFormat} from '../../Enums/PublicationItemFormat';
import { SVGIcon } from '../SVGIcon';


export class PublicationItem extends Item {
  public id: number;
  public edition: string;
  public format: PublicationItemFormat;
  public publicationItemType: PublicationItemType;
  public purchasePrice: number;
  public rentalRate: number;
  public publicationItemStatus: PublicationItemStatus;
  public publication: Publication;
  public quantity: number;
  public svgIcon:SVGIcon;


  constructor(...data: any) {
    data = data[0];
    super(data);
    if (data) {
      this.id = data["id"] || null;
      this.edition = data["edition"] || null;
      this.format = data["format"] || null;
      this.publicationItemType = data["publicationItemType"] || null;
      this.purchasePrice = data["purchasePrice"] || null;
      this.rentalRate = data["rentalRate"] || null;
      this.publicationItemStatus = data["publicationItemStatus"] || null;
      this.publication = data["publication"] || null;
      this.quantity = data["quantity"] || null;
      this.svgIcon = data["svgIcon"] || null;
    }
    this.itemType = ItemType.PUBLICATION_ITEM;
  }
}
